filebeat/harvester/log.go

```diff
@@ -186,52 +216,38 @@ func (h *Harvester) openFile() error {
 		return err
 	}
 
-	// yay, open file
-	h.file = source.File{f}
 	return nil
 }
 
 func (h *Harvester) initFileOffset(file *os.File) error {
-	offset, err := file.Seek(0, os.SEEK_CUR)
 
-	if h.getOffset() > 0 {
-		// continue from last known offset
-		logp.Debug("harvester",
-			"harvest: %q position:%d (offset snapshot:%d)", h.path, h.getOffset(), offset)
-		_, err = file.Seek(h.getOffset(), os.SEEK_SET)
-	} else if h.config.TailFiles {
-		// tail file if file is new and tail_files config is set
-		logp.Debug("harvester", "harvest: (tailing) %q (offset snapshot:%d)", h.path, offset)
+	var err error
 
-		offset, err = file.Seek(0, os.SEEK_END)
-		h.SetOffset(offset)
+	// continue from last known offset
+	if h.state.Offset > 0 {
+		h.state.Offset, err = file.Seek(h.state.Offset, os.SEEK_SET)
+		logp.Debug("harvester", "Setting offset for file: %s. Offset: %d ", h.state.Source, h.state.Offset)
+		return err
+	}
 
-	} else {
-		// get offset from file in case of encoding factory was
-		// required to read some data.
-		logp.Debug("harvester", "harvest: %q (offset snapshot:%d)", h.path, offset)
-		h.SetOffset(offset)
+	// tail file if file is new and tail_files config is set
+	if h.config.TailFiles {
+		h.state.Offset, err = file.Seek(0, os.SEEK_END)
+		logp.Debug("harvester", "Setting offset for tailing file: %s. Offset (end of file): %d ", h.state.Source, h.state.Offset)
+		return err
 	}
 
+	// get offset from file in case of encoding factory was required to read some data.
+	h.state.Offset, err = file.Seek(0, os.SEEK_CUR)
```

Comment:
Based on your input I made two adjustments:

* offset is only updated if seek does not return any error. I move the update of the offset to the validate method
* state update is only sent in the close method, if file was successfully initialised. I think it should not make a difference as the update would stay the same if no change happened, but it is not necessary and somehow makes it clear, that on setup error no updates are sent.